firestore/src/app/homepage/homepage.component.spec.ts

```diff
@@ -14,24 +14,129 @@
+
+import { DEFAULT_SORT_DATA } from '../filter-dialog/dialogdata';
+import { Observable, of } from 'rxjs';
+import { Restaurant } from 'types/restaurant';
+
+@Injectable()
+class MockHomepageFirestore extends HomepageFirestore {
+  override getRestaurantCollectionData(): Observable<Restaurant[]> {
+    const mockRestaurants: Restaurant[] = [{
+      id: "Mock 1",
+      avgRating: 3,
+      category: "Italian",
+      city: "Atlanta",
+      name: "Mock Eats 1",
+      numRatings: 0,
+      photo: "Mock Photo URL",
+      price: 1
+    }]
+
+    return of(mockRestaurants);
+  }
+
+  override getRestaurntsGivenConstraints(constraints: QueryConstraint[]): Observable<Restaurant[]> {
+    const mockRestaurants: Restaurant[] = [{
+      id: "Mock 1",
+      avgRating: 3,
+      category: "Italian",
+      city: "Atlanta",
+      name: "Mock Eats 1",
+      numRatings: 0,
+      photo: "Mock Photo URL",
+      price: 1
+    },
+    {
+      id: "Mock 2",
+      avgRating: 3,
+      category: "Korean",
+      city: "Los Angeles",
+      name: "Mock Eats 2",
+      numRatings: 0,
+      photo: "Mock Photo URL",
+      price: 2
+    }]
+
+    return of(mockRestaurants);
+  }
+}
+
 
 describe('HomepageComponent', () => {
   let component: HomepageComponent;
   let fixture: ComponentFixture<HomepageComponent>;
-
+  let mockDialog = jasmine.createSpyObj('MatDialog', ['open']);
   beforeEach(() => {
     TestBed.configureTestingModule({
-      declarations: [HomepageComponent]
+      imports: [MatDialogModule,
+        MatToolbarModule,
+        MatIconModule,
+        provideFirebaseApp(() => initializeApp(projectConfig)),
+        provideFirestore(() => getFirestore()),
+        provideAuth(() => getAuth())],
+      declarations: [HomepageComponent],
+      providers: [
+        {
+          provide: HomepageFirestore,
+          useClass: MockHomepageFirestore
+        },
+        { provide: MAT_DIALOG_DATA, useValue: {} },
+        { provide: MatDialogRef, useValue: {} }
+      ],
     });
     fixture = TestBed.createComponent(HomepageComponent);
     component = fixture.componentInstance;
+    component.dialog = mockDialog;
     fixture.detectChanges();
   });
 
   it('should create', () => {
     expect(component).toBeTruthy();
   });
-});
+
+  /** 
+   * Tests to see that the #empty-restaurants-container div has not
+   * been rendered. This div is rendered only when no restaurants have been 
+   * retrieved from Firestore (or, in this case, the MockHomepageFirestore 
+   * element).
+   * */
+  it('should call service and get restuarants on init', () => {
+    const emptyRestaurantsDiv = fixture.debugElement.query(By.css("#empty-restaurants-container"));
+    expect(emptyRestaurantsDiv).toBeNull();
+  });
+
+  it('should get new data when filters change', waitForAsync(() => {
+    let mockDialogRef = jasmine.createSpyObj('MatDialogRef', ['afterClosed']);
+    mockDialogRef.afterClosed.and.returnValue(of({ ...DEFAULT_SORT_DATA, price: 2 }));
```

Comment:
Mock dialog works here if we want to focus testing on the homepage component itself